File: Ecosystem/Creature.m
classdef Creature
    % Base creature class to be used for predators and herbivores
    
    properties
        type = "none";
        food_type = "none";
        enemy_type = "none";
        action = "none";
        % Status Levels
        stress_level = 0;
        hunger_level = 0.5;
        health_level = 1;
        age = 0;
        mature = 15;
        sight_range = 3;
        times_reproduced = 0;
        max_reproduce = 3;
        regen = 0.05;
        damage = 0.1;
        nourish = 0.1;
        alive = true;
        
        % Weights
        stress_weight = 0.5;
        hunger_weight = 0.5;
        health_weight = 0.5;
        health_thresh = 0.5;
        hunger_thresh = 0.5;
        stress_thresh = 0.5;
        % Personality Makeup (1 -> 0, first term -> second term)
        aggressive_placid = 0.5;
        fear_brave = 0.5;
        social_isolationist = 0.5;
        
        % Location
        location = [];
        friendlies_close = {};
        enemies_close = {};
        food_close = {};
        friend_distances = [];
        enemy_distances = [];
        food_distances = [];
        other_close = {};
        closest_enemy = struct();
        closest_friend = struct();
        closest_food = struct();
        enemy_loc_index = 0;
        friend_loc_index = 0;
        food_loc_index = 0;
        friend_location = [];
        enemy_location = [];
        food_location = [];
        friendlies_indices =[];
        enemies_indices = [];
        food_indices = [];
        
        % List of unformatted objects
        entities_close = {};
        entities_dist= [];
        child_location = [];
        map_size = 0;
        
    end
    
    methods
        
        % Look
        function obj = process_sight(obj,object_list,map_size)
            obj.map_size = map_size;
            af_fr = false;
            af_e = false;
            af_fd = false;
            af_o = false;
            obj.friendlies_close = {};
            obj.friendlies_indices = [];
            obj.enemies_close = {};
            obj.enemies_indices = [];
            obj.food_close = {};
            obj.food_indices = [];
            
            % Sort through list of known objects
            temp_object_locations = cellfun(@(x) x.location, object_list,...
                                            "UniformOutput", false);
            temp_object_locations = vertcat(temp_object_locations{:});
            object_dist = [];
            
            for i = 1:size(temp_object_locations,1)
                object_dist(end+1) = pdist2(obj.location,temp_object_locations(i,:));
            end
            
            close_objects_ind = find(object_dist <= obj.sight_range);
            close_objects_list = object_list(close_objects_ind);
            obj.entities_close = close_objects_list;
            obj.entities_dist = object_dist(close_objects_ind);
            
            
            for i = 1:numel(close_objects_list)
                
                inspected_object = close_objects_list{i};
                creature_xy = inspected_object.location;
                relative_distance = pdist2(obj.location,creature_xy);
                
                if relative_distance <= obj.sight_range
                    
                    switch inspected_object.type
                        
                        case obj.type
                            switch af_fr
                                case true
                                    obj.friendlies_close(end+1) = close_objects_list(i);
                                    obj.friendlies_indices(end + 1) = i;
                                case false
                                    obj.friendlies_close(1) = close_objects_list(i);
                                    obj.friendlies_indices(1) = i;
                                    af_fr = true;
                            end
                            
                        case obj.food_type
                            switch af_fd
                                case true
                                    obj.food_close(end+1) = close_objects_list(i);
                                    obj.food_indices(end + 1) = i;
                                case false
                                    obj.food_close(1) = close_objects_list(i);
                                    obj.food_indices(1) = i;
                                    af_fd = true;
                            end
                            
                        case obj.enemy_type
                            switch af_e
                                case true
                                    obj.enemies_close(end+1) = close_objects_list(i);
                                    obj.enemies_indices(end + 1) = i;
                                case false
                                    obj.enemies_close(1) = close_objects_list(i);
                                    obj.enemies_indices(1) = i;
                                    af_e = true;
                            end
                            
                        otherwise
                            switch af_o
                                case true
                                    obj.other_close(end+1) = close_objects_list(i);
                                case false
                                    obj.other_close(1) = close_objects_list(i);
                                    af_o = true;
                            end  
                    end
                    
                end
            end
            
            % Convert to distances
            % enemies
            for i = 1:numel(obj.enemies_close)
                yx = obj.enemies_close{i}.location;
                y = yx(1);
                x = yx(2);
                obj.enemy_distances(i) = pdist2(obj.location,[y x]);
            end
            
            obj.enemy_loc_index = find(obj.enemy_distances == min(obj.enemy_distances));
           
            if numel(obj.enemy_loc_index) ~=0 && (numel(obj.enemies_close) ~= 0)
                obj.enemy_loc_index = obj.enemy_loc_index(1);
                obj.closest_enemy = obj.enemies_close{obj.enemy_loc_index};
                obj.enemy_location = obj.closest_enemy.location;
                obj.enemy_loc_index = obj.enemies_indices(obj.enemy_loc_index);
            end
            
            
            % friends
            for i = 1:numel(obj.friendlies_close)
                yx = obj.friendlies_close{i}.location;
                y = yx(1);
                x = yx(2);
                obj.friend_distances(i) = pdist2(obj.location,[y x]);
            end
            
            obj.friend_loc_index = find(obj.friend_distances == min(obj.friend_distances));
            
            if numel(obj.friend_loc_index) ~=0 && (numel(obj.friendlies_close) ~= 0)
                obj.friend_loc_index = obj.friend_loc_index(1);
                obj.closest_friend = obj.friendlies_close{obj.friend_loc_index};
                obj.friend_location = obj.closest_friend.location;
                obj.friend_loc_index = obj.friendlies_indices(obj.friend_loc_index);
            end
            
            %food
            for i = 1:numel(obj.food_close)
                yx = obj.food_close{i}.location;
                y = yx(1);
                x = yx(2);
                obj.food_distances(i) = pdist2(obj.location,[y x]);
            end

            obj.food_loc_index = find(obj.food_distances == min(obj.food_distances));
            
            if numel(obj.food_loc_index) ~=0 && (numel(obj.food_close) ~= 0)
                obj.food_loc_index = obj.food_loc_index(1);
                obj.closest_food = obj.food_close{obj.food_loc_index};
                obj.food_location = obj.closest_food.location;
                obj.food_loc_index = obj.food_indices(obj.food_loc_index);
            end
            
        end
        
        % Process Stress
        function obj = process_stress(obj)
            % Calculate stress level
            stress_base = 0.25;
            amount_enemies = numel(obj.enemies_close);
            amount_friends = numel(obj.friendlies_close);
            amount_food = numel(obj.food_close);
            
            % Stress level due to threat
            threat_weight = amount_enemies/((obj.sight_range*2)^2);
            threat_stress = (threat_weight)*stress_base;
            
            % Stress level due to friends
            friend_weight = amount_friends/((obj.sight_range*2)^2);
            friend_stress = (friend_weight*stress_base)+(obj.social_isolationist*(stress_base/2));
            if friend_stress >stress_base
                friend_stress = stress_base;
            end
            
            % Stress level due to food
            food_stress = (obj.hunger_level*(stress_base/2))+(obj.hunger_weight*(stress_base/2));
            
            % Stress level due to health
            health_stress = (obj.health_level*(stress_base/2))+(obj.health_weight*(stress_base/2));
            
            % Stress level
            obj.stress_level = (threat_stress + food_stress + health_stress) - friend_stress;
            
            if obj.stress_level < 0
                obj.stress_level = 0;
                
            elseif obj.stress_level >1
                obj.stress_level = 1;
                
            end
        end
        
        % Think
        function obj = process_thought(obj)
            obj.health_thresh = obj.health_weight * obj.health_level;
            obj.hunger_thresh = obj.hunger_weight * obj.hunger_level;
            obj.stress_thresh = obj.stress_weight * obj.stress_level;
            
            is_mature = obj.age > obj.mature;
            is_full = obj.hunger_level > obj.hunger_thresh;
            is_healthy = obj.health_level > obj.health_thresh;
            is_stressed = obj.stress_level > obj.stress_thresh;
            enemies_near = numel(obj.enemies_close) > 0;
            friends_near = numel(obj.friendlies_close) > 0;
            food_near = numel(obj.food_close) > 0;
            out_numbered = numel(obj.enemies_close) > (numel(obj.friendlies_close)+1);
            
            % Reproduce
            if(is_mature && is_healthy && is_full && ~is_stressed && ...
                    ~enemies_near && friends_near)
                obj.action = "rep";
                
                
                % Rest
            elseif((~is_healthy) && (is_full))
                obj.action = "rest";
                
                
                % Eat
            elseif ~is_full && (all(min(obj.enemy_distances) > 2)) && food_near
                obj.action = "eat";
                % Attack
            elseif(enemies_near && ~is_stressed && is_healthy)
                obj.action = "attack";
                % Flee
            elseif(enemies_near && (is_stressed || ~is_healthy || ...
                    out_numbered))
                obj.action = "flee";
                % Wander
            else
                obj.action = "wander";
            end
            
        end
        
        % Act
        function [obj, recent_action] = process_action(obj)
            
            recent_action = "nothing";
            
            % Which action to select
            switch obj.action
                
                case "rep"
                    %Is friend near
                    distance = obj.nearest_creature(obj,obj.type);
                    if distance == 1
                        o_location = obj.find_empty(obj);
                        if numel(o_location) ~= 0
                            obj.child_location = o_location;
                            recent_action = "rep";

                        else
                            obj.child_location = [];
                        end

                    % Friend not near
                    elseif distance > 1
                        o_location = obj.find_near_empty(obj, obj.closest_friend.location);

                        if numel(o_location) ~= 0
                           obj.location = o_location;
                        end
                    end
                    
                case "rest"
                    
                    if obj.health_level ~= 1
                        obj = regenerate(obj);
                    end
                    
                case "eat"
                    % Food near
                    distance = obj.nearest_creature(obj,obj.food_type);
                    if distance == 1
                        recent_action = "eat";
                        obj.hunger_level = obj.hunger_level + obj.nourish;

                    % Food not near
                    elseif distance > 1
                        o_location = obj.find_near_empty(obj, obj.closest_food.location);

                        if numel(o_location) ~= 0
                           obj.location = o_location;
                        end
                    end 
                    
                case "attack"
                    
                    % Enemy near
                    distance = obj.nearest_creature(obj,obj.enemy_type);
                    if distance == 1
                        recent_action = "attack";
                        
                    % Enemy not near
                    elseif distance > 1
                        o_location = obj.find_near_empty(obj, obj.closest_enemy.location);

                        if numel(o_location) ~= 0
                           obj.location = o_location;
                        end
                    end                     
                    
                case "flee"
                    o_location = obj.find_near_empty(obj, obj.closest_enemy.location, "flee");
                    
                    if numel(o_location) ~= 0
                        obj.location = o_location;
                    end
                    
                case "wander"
                    o_location = obj.find_empty(obj, 0, true);
                    
                    if numel(o_location) ~= 0
                        obj.location = o_location;
                    end
            end
        end
        
        % Regen
        function obj = regenerate(obj)
            obj.health_level = obj.health_level + obj.regen;
        end
        
        % Hurt
        function obj = hurt(obj)
           obj.health_level = obj.health_level - obj.damage;
        end
    end
    
    methods(Static)
        
        function distance = nearest_creature(object,type)
            % Function for finding nearest POI
            
            % Finding the nearest one's euclidean distance
            objects_near = object.entities_close;
            distance = inf;
            
            for i = 1:numel(objects_near)
                
                if objects_near{i}.type == type
                    iteration_distance = pdist2(object.location,objects_near{i}.location);
                    
                    if iteration_distance < distance
                        distance = iteration_distance;
                        
                    end
                end
            end
            
            if distance == inf
                distance = 0;
            end
            
        end

        function location = find_empty(object, iteration_skip, random_sel)
            % Function for finding the nearest empty tile
            if nargin < 2
                iteration_skip = 0;
                random_sel = false;
            
            elseif nargin < 3
                random_sel = false;
            end
            
            map_size = object.map_size;
            
            location = [];
            near_objects = object.entities_close;
            
            % Create array of possible areas to check
            locations_check = [];
            count = 1;
            for i = -1:1
                
                for k = -1:1
                    
                    if (i == 0) && (k == 0)
                        continue
                    end
                    yx = object.location;
                    y = yx(1);
                    x = yx(2);
                    n_y = y + i;
                    n_x = x + k;
                    if (n_y > object.map_size) || (n_x > object.map_size)...
                            || (n_y) < 1 || n_x < 1
                        continue
                    end
                    locations_check(count,:) = [n_y n_x]; %#ok<*AGROW>
                    count = count + 1;
                    
                end
                
            end
            
            if random_sel == true
                locations_check = locations_check(randperm(size(locations_check, 1)), :);
            end
            % Iterate through locations to check and iterate through locations
            for i = 1:length(locations_check)
                
                for k = 1:numel(near_objects)
                    logical_loc = near_objects{k}.location == locations_check(i,:);
                    if (length(map_size) < length(locations_check(i)))
                       continue 
                    end
                        
                    if ~all(logical_loc)
                        location = locations_check(i, :);
                        % Used to look for different spot
                        if iteration_skip > 0
                            iteration_skip = iteration_skip - 1;
                            location = [];
                            continue
                        end
                        
                        break
                        
                    end
                end
            end

       end

        function location = find_near_empty(object, POI_location, pursuit)
            % Function for finding closest movement position

            % Default final argument
            if nargin < 3
               pursuit = "pursue"; 
            end
            location = object.location;
            distance = pdist2(object.location,POI_location);

            for i = 0:7 
                checked_location = object.find_empty(object,i);

                if numel(checked_location) == 0
                   continue 
                end

                checked_distance = pdist2(checked_location,POI_location);

                switch pursuit

                    case "pursue"
                        if checked_distance < distance && checked_distance ~= 0
                           location = checked_location; 
                        end

                    case "flee"
                        if checked_distance > distance
                            location = checked_location;
                        end
                end
            end

        end
        
    end
end

File: Ecosystem_Python/creature.py
import numpy as np
from typing import List, Tuple, Optional, Dict, Any
import random


class Creature:
    """Base creature class to be used for predators and herbivores"""
    
    def __init__(self, location: List[int]):
        # Basic properties
        self.type = "none"
        self.food_type = "none"
        self.enemy_type = "none"
        self.action = "none"
        
        # Status Levels
        self.stress_level = 0.0
        self.hunger_level = 0.5
        self.health_level = 1.0
        self.age = 0
        self.mature = 15
        self.sight_range = 3
        self.times_reproduced = 0
        self.max_reproduce = 3
        self.regen = 0.05
        self.damage = 0.1
        self.nourish = 0.1
        self.alive = True
        
        # Weights
        self.stress_weight = 0.5
        self.hunger_weight = 0.5
        self.health_weight = 0.5
        self.health_thresh = 0.5
        self.hunger_thresh = 0.5
        self.stress_thresh = 0.5
        
        # Personality Makeup (1 -> 0, first term -> second term)
        self.aggressive_placid = 0.5
        self.fear_brave = 0.5
        self.social_isolationist = 0.5
        
        # Location and nearby objects
        self.location = np.array(location)
        self.friendlies_close = []
        self.enemies_close = []
        self.food_close = []
        self.friend_distances = []
        self.enemy_distances = []
        self.food_distances = []
        self.other_close = []
        self.closest_enemy = None
        self.closest_friend = None
        self.closest_food = None
        self.enemy_loc_index = 0
        self.friend_loc_index = 0
        self.food_loc_index = 0
        self.friend_location = []
        self.enemy_location = []
        self.food_location = []
        self.friendlies_indices = []
        self.enemies_indices = []
        self.food_indices = []
        
        # List of unformatted objects
        self.entities_close = []
        self.entities_dist = []
        self.child_location = []
        self.map_size = 0
    
    def process_sight(self, object_list: List['Creature'], map_size: int) -> 'Creature':
        """Process sight to identify nearby objects"""
        self.map_size = map_size
        
        # Reset sight data
        self.friendlies_close = []
        self.friendlies_indices = []
        self.enemies_close = []
        self.enemies_indices = []
        self.food_close = []
        self.food_indices = []
        
        # Get all object locations
        object_locations = np.array([obj.location for obj in object_list])
        
        # Calculate distances to all objects
        distances = np.linalg.norm(object_locations - self.location, axis=1)
        
        # Find objects within sight range
        close_indices = np.where(distances <= self.sight_range)[0]
        close_objects = [object_list[i] for i in close_indices]
        self.entities_close = close_objects
        self.entities_dist = distances[close_indices]
        
        # Categorize nearby objects
        for i, obj in enumerate(close_objects):
            global_index = close_indices[i]
            
            if obj.type == self.type:
                self.friendlies_close.append(obj)
                self.friendlies_indices.append(global_index)
            elif obj.type == self.food_type:
                self.food_close.append(obj)
                self.food_indices.append(global_index)
            elif obj.type == self.enemy_type:
                self.enemies_close.append(obj)
                self.enemies_indices.append(global_index)
            else:
                self.other_close.append(obj)
        
        # Calculate distances to each category
        self._calculate_category_distances()
        
        return self
    
    def _calculate_category_distances(self):
        """Calculate distances to enemies, friends, and food"""
        # Enemy distances
        self.enemy_distances = []
        if self.enemies_close:
            self.enemy_distances = [np.linalg.norm(enemy.location - self.location) 
                                  for enemy in self.enemies_close]
            min_enemy_idx = np.argmin(self.enemy_distances)
            self.closest_enemy = self.enemies_close[min_enemy_idx]
            self.enemy_location = self.closest_enemy.location
            self.enemy_loc_index = self.enemies_indices[min_enemy_idx]
        
        # Friend distances
        self.friend_distances = []
        if self.friendlies_close:
            self.friend_distances = [np.linalg.norm(friend.location - self.location) 
                                   for friend in self.friendlies_close]
            min_friend_idx = np.argmin(self.friend_distances)
            self.closest_friend = self.friendlies_close[min_friend_idx]
            self.friend_location = self.closest_friend.location
            self.friend_loc_index = self.friendlies_indices[min_friend_idx]
        
        # Food distances
        self.food_distances = []
        if self.food_close:
            self.food_distances = [np.linalg.norm(food.location - self.location) 
                                 for food in self.food_close]
            min_food_idx = np.argmin(self.food_distances)
            self.closest_food = self.food_close[min_food_idx]
            self.food_location = self.closest_food.location
            self.food_loc_index = self.food_indices[min_food_idx]
    
    def process_stress(self) -> 'Creature':
        """Calculate stress level based on environment"""
        stress_base = 0.25
        amount_enemies = len(self.enemies_close)
        amount_friends = len(self.friendlies_close)
        
        # Stress level due to threat
        threat_weight = amount_enemies / ((self.sight_range * 2) ** 2)
        threat_stress = threat_weight * stress_base
        
        # Stress level due to friends
        friend_weight = amount_friends / ((self.sight_range * 2) ** 2)
        friend_stress = (friend_weight * stress_base + 
                        self.social_isolationist * (stress_base / 2))
        friend_stress = min(friend_stress, stress_base)
        
        # Stress level due to food
        food_stress = (self.hunger_level * (stress_base / 2) + 
                      self.hunger_weight * (stress_base / 2))
        
        # Stress level due to health
        health_stress = (self.health_level * (stress_base / 2) + 
                        self.health_weight * (stress_base / 2))
        
        # Calculate total stress
        self.stress_level = (threat_stress + food_stress + health_stress) - friend_stress
        self.stress_level = max(0, min(1, self.stress_level))
        
        return self
    
    def process_thought(self) -> 'Creature':
        """Process decision making based on current state"""
        self.health_thresh = self.health_weight * self.health_level
        self.hunger_thresh = self.hunger_weight * self.hunger_level
        self.stress_thresh = self.stress_weight * self.stress_level
        
        is_mature = self.age > self.mature
        is_full = self.hunger_level > self.hunger_thresh
        is_healthy = self.health_level > self.health_thresh
        is_stressed = self.stress_level > self.stress_thresh
        enemies_near = len(self.enemies_close) > 0
        friends_near = len(self.friendlies_close) > 0
        food_near = len(self.food_close) > 0
        out_numbered = len(self.enemies_close) > (len(self.friendlies_close) + 1)
        
        # Decision making hierarchy
        if (is_mature and is_healthy and is_full and not is_stressed and 
            not enemies_near and friends_near):
            self.action = "rep"
        elif not is_healthy and is_full:
            self.action = "rest"
        elif (not is_full and food_near and 
              (not self.enemy_distances or min(self.enemy_distances) > 2)):
            self.action = "eat"
        elif enemies_near and not is_stressed and is_healthy:
            self.action = "attack"
        elif enemies_near and (is_stressed or not is_healthy or out_numbered):
            self.action = "flee"
        else:
            self.action = "wander"
        
        return self
    
    def process_action(self) -> Tuple['Creature', str]:
        """Execute the chosen action"""
        recent_action = "nothing"
        
        if self.action == "rep":
            distance = self._nearest_creature_distance(self.type)
            if distance == 1:
                o_location = self._find_empty()
                if o_location is not None:
                    self.child_location = o_location
                    recent_action = "rep"
                else:
                    self.child_location = None
            elif distance > 1 and self.closest_friend is not None:
                o_location = self._find_near_empty(self.closest_friend.location)
                if o_location is not None:
                    self.location = o_location
        
        elif self.action == "rest":
            if self.health_level != 1:
                self.regenerate()
        
        elif self.action == "eat":
            distance = self._nearest_creature_distance(self.food_type)
            if distance == 1:
                recent_action = "eat"
                self.hunger_level = min(1.0, self.hunger_level + self.nourish)
            elif distance > 1 and self.closest_food is not None:
                o_location = self._find_near_empty(self.closest_food.location)
                if o_location is not None:
                    self.location = o_location
        
        elif self.action == "attack":
            distance = self._nearest_creature_distance(self.enemy_type)
            if distance == 1:
                recent_action = "attack"
            elif distance > 1 and self.closest_enemy is not None:
                o_location = self._find_near_empty(self.closest_enemy.location)
                if o_location is not None:
                    self.location = o_location
        
        elif self.action == "flee":
            if self.closest_enemy is not None:
                o_location = self._find_near_empty(self.closest_enemy.location, "flee")
                if o_location is not None:
                    self.location = o_location
        
        elif self.action == "wander":
            o_location = self._find_empty(random_sel=True)
            if o_location is not None:
                self.location = o_location
        
        return self, recent_action
    
    def regenerate(self) -> 'Creature':
        """Regenerate health"""
        self.health_level = min(1.0, self.health_level + self.regen)
        return self
    
    def hurt(self) -> 'Creature':
        """Take damage"""
        self.health_level = max(0.0, self.health_level - self.damage)
        return self
    
    def _nearest_creature_distance(self, creature_type: str) -> float:
        """Find distance to nearest creature of specified type"""
        min_distance = float('inf')
        
        for obj in self.entities_close:
            if obj.type == creature_type:
                distance = np.linalg.norm(obj.location - self.location)
                min_distance = min(min_distance, distance)
        
        return 0 if min_distance == float('inf') else min_distance
    
    def _find_empty(self, iteration_skip: int = 0, random_sel: bool = False) -> Optional[np.ndarray]:
        """Find nearest empty tile"""
        locations_check = []
        
        # Generate adjacent locations
        for i in range(-1, 2):
            for k in range(-1, 2):
                if i == 0 and k == 0:
                    continue
                
                new_y = self.location[0] + i
                new_x = self.location[1] + k
                
                if (new_y > self.map_size or new_x > self.map_size or 
                    new_y < 1 or new_x < 1):
                    continue
                
                locations_check.append([new_y, new_x])
        
        if random_sel and locations_check:
            random.shuffle(locations_check)
        
        # Check each location for emptiness
        for location in locations_check:
            is_empty = True
            for obj in self.entities_close:
                if np.array_equal(obj.location, location):
                    is_empty = False
                    break
            
            if is_empty:
                if iteration_skip > 0:
                    iteration_skip -= 1
                    continue
                return np.array(location)
        
        return None
    
    def _find_near_empty(self, poi_location: np.ndarray, pursuit: str = "pursue") -> Optional[np.ndarray]:
        """Find closest movement position relative to point of interest"""
        best_location = self.location.copy()
        current_distance = np.linalg.norm(self.location - poi_location)
        
        for i in range(8):
            checked_location = self._find_empty(iteration_skip=i)
            
            if checked_location is None:
                continue
            
            checked_distance = np.linalg.norm(checked_location - poi_location)
            
            if pursuit == "pursue":
                if checked_distance < current_distance and checked_distance != 0:
                    best_location = checked_location
                    current_distance = checked_distance
            elif pursuit == "flee":
                if checked_distance > current_distance:
                    best_location = checked_location
                    current_distance = checked_distance
        
        return best_location if not np.array_equal(best_location, self.location) else None